Draft a blog post about integrating Kedro and Ibis (#127)
kedro-org/kedro-devrel: blog_post_collaboration/kedro_ibis.md

# From production-ready to production: building scalable data pipelines with Kedro and Ibis

In your Kedro journey, have you ever...

> * ...slurped up large amounts of data into memory, instead of pushing execution down to the source database/engine?

Reviewer: Nice! Great way to get the audience to relate to the content!

> * ...implemented a proof-of-concept solution in 3-4 months on data extracts, and then struggled massively when you needed to move to running against the production databases and scale out?

Reviewer 2: kind of related, not sure how to phrase, but similarly "prototypes on pandas and had to convert code over to PySpark/Snowpark/other database-native Python dataframes?"

Reviewer 2: reading on this is already addressed; perhaps a bullet point up here? but 🤷

In practice, a machine learning (software) engineer often ends up rewriting proof-of-concept code to be more performant and scalable, but this presents new challenges. First, it takes time to reimplement everything properly in another technology, and developers frequently underestimate the effort required. Also, rewritten logic, intentionally or not, doesn't always correspond 1:1 to the original, leading to unexpected results (read: bugs). Proper testing during development can mitigate, but not eliminate, correctness issues, but expecting all data science code to be unit tested is wishful thinking. Last but not least, business stakeholders don't always understand why productionising code takes so much time and resources—after all, it worked before! Who knows? Maybe scaling data engineering code shouldn't be so hard...

> Given the ubiquity of pandas during the development phase, some execution engines started supporting pandas-compatible or similar Python dataframe APIs, such as the [pandas API on Spark](https://spark.apache.org/docs/latest/api/python/user_guide/pandas_on_spark/index.html) and [BigQuery DataFrames](https://cloud.google.com/bigquery/docs/reference/bigquery-dataframes). Case studies, like [this one from Virgin Hyperloop

Reviewer 2: might be worth mentioning the BigQuery pandas API is built on Ibis

One](https://www.databricks.com/blog/2019/08/22/guest-blog-how-virgin-hyperloop-one-reduced-processing-time-from-hours-to-minutes-with-koalas.html), demonstrate how these solutions can accelerate the scaling process for data-intensive code. That said, building and maintaining a pandas-compatible API requires significant commitment from the engine developer, so no panacea exists.

## The SQL solution

Thankfully, there exists a standardized programming language that every database (and more-or-less every major computation framework) supports: SQL![^1]

> _If you like SQL,_ [dbt](https://www.getdbt.com) provides a battle-tested framework for defining and deploying data transformation workflows. [As of February 2022, over 9,000 companies were using dbt in production.](https://www.getdbt.com/blog/next-layer-of-the-modern-data-stack) Here on the Kedro team, we love dbt! In fact, we frequently recommend it for ELT (Extract, Load, Transform) workflows. Kedro and dbt share similar self-stated goals: [Kedro empowers data teams to write reproducible, maintainable, and modular Python code,](https://docs.kedro.org/en/stable/introduction/index.html#introduction-to-kedro) and ["dbt is a SQL-first transformation workflow that lets teams quickly and collaboratively deploy analytics code following software engineering best practices."](https://www.getdbt.com/product/what-is-dbt)

Reviewer 3: (Suggestion: Split the text in several lines to aid reviews)

Reviewer 3: Worth clarifying "for the T in ELT workflows"? Theoretically, Kedro can be used for EL, which is outside the scope of dbt (whether this is optimal or not is a separate debate)

Author: Good catch!

> _"When I learned about Kedro (while at dbt Labs), I commented that it was like dbt if it were created by Python data scientists instead of SQL data analysts (including both being created out of consulting companies)."_
>

> > —Cody Peterson, Technical Product Manager

Reviewer 2: might want to add " @ Voltron Data" just to be clear I don't work at dbt Labs anymore

## "What if I don't want to use SQL?"

This post isn't going to enter the SQL vs. Python debate. Nevertheless, many teams and data scientists choose Python for their analytics workflows. Furthermore, in my experience, teams using Kedro for their machine learning pipelines prefer to use the same tool for data processing and feature engineering code, too.

> Can we combine the flexibility and familiarity of Python with the scale and performance of modern SQL? Yes we can! ["Ibis is a Python library that provides a lightweight, universal interface for data wrangling."](https://github.com/ibis-project/ibis) Ibis exposes a Python dataframe API where the same code can execute against over 15 query engines, from local backends like pandas, Polars, and DuckDB to remote databases and distributed computation frameworks like BigQuery, Snowflake, and Spark. Ibis constructs a query plan, or intermediate representation (IR), that it evaluates lazily (i.e. as and when needed) on the execution engine.

Reviewer 2: might be worth mentioning that Ibis also makes it easy to use SQL strings in the same codebase with `.sql` and `.to_sql` methods -- i.e. you're not losing the ability to leverage SQL

## Reimagining Python-first production data pipelines

> ### Revisiting the Jaffle Shop

Reviewer: ❤️

Since dbt represents the industry standard for SQL-first transformation workflows, let's use the Jaffle Shop example to inform the capabilities of a Python-first solution. For those unfamiliar with the Jaffle Shop project, it provides a playground dbt project for testing and demonstration purposes, akin to the Kedro spaceflights project. Take a few minutes to [try dbt with DuckDB yourself](https://github.com/dbt-labs/jaffle_shop_duckdb) or [watch a brief walkthrough of the same](https://www.loom.com/share/ed4a6f59957e43158837eb4ba0c5ed67).

If you want to peek at the final solution using Kedro at this point, [jump to the "Quickstart" section](#quickstart).

### Creating a custom `ibis.Table` dataset

> For our Kedro-Ibis integration, we need the ability to load and save data using the backends Ibis supports. The data will be represented in memory as an [Ibis table](https://ibis-project.org/reference/expression-tables#ibis.expr.types.relations.Table), analagous to a pandas dataframe. The first step in using Ibis involves connecting to the desired backend; for example, see [how to connect to the DuckDB backend](https://ibis-project.org/backends/duckdb#connect). We abstract this step in the dataset configuration. We also maintain a mapping of established connections for reuse—a technique borrowed from existing dataset implementations, like [those of `pandas.SQLTableDataset` and `pandas.SQLQueryDataset`](https://github.com/kedro-org/kedro-plugins/blob/kedro-datasets-2.0.0/kedro-datasets/kedro_datasets/pandas/sql_dataset.py).

Reviewer 3: s/analagous/analogous/g

Reviewer 3: Also, AFAIK an Ibis table is lazy, rather than materialized? Hence it's not exactly "analogous" (worth picking a better word, given that the lazy nature was mentioned in the Ibis introduction paragraph)

Author: Call me lazy, but I just added "(lazy)". :D

We also need to support different [materialisation strategies](https://docs.getdbt.com/docs/build/materializations). Depending on what the user configures, we call either [`create_table`](https://ibis-project.org/backends/duckdb#ibis.backends.duckdb.Backend.create_table) or [`create_view`](https://ibis-project.org/backends/duckdb#ibis.backends.duckdb.Backend.create_view).

> [Find the complete dataset implemention on GitHub.](https://github.com/deepyaman/jaffle-shop/blob/main/src/jaffle_shop/datasets/ibis/table_dataset.py)

Reviewer 3: nit:

```suggestion
[Find the complete dataset implemention on GitHub](https://github.com/deepyaman/jaffle-shop/blob/main/src/jaffle_shop/datasets/ibis/table_dataset.py).
```

### Configuring backends with the `OmegaConfigLoader` using variable interpolation

Since Kedro 0.18.10, you can [leverage OmegaConf-native templating in catalog files](https://docs.kedro.org/en/latest/configuration/advanced_configuration.html#catalog). In our example:

```yaml
# conf/base/catalog_connections.yml
_duckdb:
  backend: duckdb
  # `database` and `threads` are parameters for `ibis.duckdb.connect()`.
  database: jaffle_shop.duckdb
  threads: 24

# conf/base/catalog.yml
raw_customers:
  type: jaffle_shop.datasets.ibis.TableDataset
  table_name: raw_customers
  connection: ${_duckdb}
  save_args:
    materialized: table

...
```

### Building pipelines

> Develop pipelines the same way you would any other Kedro pipeline. Below, we show a node written using Ibis alongside the original dbt model. Note that Ibis solves the [parametrisation problem](https://www.youtube.com/watch?v=XdZklxTbCEA&t=514s) with vanilla Python syntax; on the other hand, the dbt counterpart uses Jinja templating.

Reviewer 3: ```suggestion
Develop pipelines the same way you would any other Kedro pipeline. Below, you can see a node written using Ibis alongside the original dbt model. Note that Ibis solves the [parametrisation problem](https://www.youtube.com/watch?v=XdZklxTbCEA&t=514s) with vanilla Python syntax; on the other hand, the dbt counterpart uses Jinja templating.
```

better to avoid first person when possible? cc @stichbury (I always guide myself by https://ddbeck.com/first-person-plural-is-questionable/)

Author: Will wait for Jo's input on this, since I probably need to change it more than just this place then.

Reviewer 4: I don't have a particular view on first person plural, and personally we think "we" is fine if used sparingly :)

But it's a bit weird in that some of this piece uses "we" and some uses "you", viz these two sentences right here: 
> Develop pipelines the same way you would any other Kedro pipeline. Below, we show a node written using Ibis alongside the original dbt model.

I would tend to remove some of the "you" in this case, so rewrite as follows:

> Develop pipelines the same way you would any other Kedro pipeline. The code below shows a node written using Ibis alongside the original dbt model.

You don't need to contort all the text to remove "you" but perhaps have a scan through the middle method section and attempt to eliminate that, keeping the you in the top and tail, which address the reader more directly, and retaining just "we" for the text that shows the steps? I can have a go if you like, but I'll let you have first dibs.

Author: > I can have a go if you like, but I'll let you have first dibs.

You're welcome to have a go! If you haven't gotten to it yet, I'll try later today or tomorrow, but I trust whatever you do will be at least as good (and almost certainly better). :)

            order_payments[f"{payment_method}_amount"]
            for payment_method in payment_methods
        ],
        order_payments.total_amount.name("amount"),
    ]
    return final
```

  <td>

```sql
{% set payment_methods = ['credit_card', 'coupon', 'bank_transfer', 'gift_card'] %}

with orders as (

    select * from {{ ref('stg_orders') }}

),

payments as (

    select * from {{ ref('stg_payments') }}

),

order_payments as (

    select
        order_id,

        {% for payment_method in payment_methods -%}
        sum(case when payment_method = '{{ payment_method }}' then amount else 0 end) as {{ payment_method }}_amount,
        {% endfor -%}

        sum(amount) as total_amount

    from payments

    group by order_id

),

final as (

    select
        orders.order_id,
        orders.customer_id,
        orders.order_date,
        orders.status,

        {% for payment_method in payment_methods -%}

        order_payments.{{ payment_method }}_amount,

        {% endfor -%}

        order_payments.total_amount as amount

    from orders

    left join order_payments
        on orders.order_id = order_payments.order_id

)

select * from final
```

</table>

> ### Quickstart

Reviewer 3: Weird to see "Quickstart" almost at the end of the blog post. Maybe "See it for yourself"?

Have you ever found yourself developing data pipelines against an existing data warehouse, even though you know that the data infrastructure team is currently migrating to a different database solution? You expect to rewrite some amount of your code—unless you use Ibis?

Leveraging Ibis can also help you build truly reusable pipelines. I previously led development of a Kedro-based code asset that provided a suite of reusable pipelines for customer analytics. We decided to use PySpark for data engineering, but many of our users didn't need to use PySpark. One major retailer stored all of their data in Oracle, and they ended up with a suboptimal workflow wherein they extracted data into Spark and did all the work there. Other teams looked at our pipelines for inspiration, but ended up rewriting them for their infrastructure—not our intention in delivering prebuilt pipelines. Last but not least, we spent so much time and effort setting up Spark on locked-down Windows machines, so data scientists could play with the pipelines we provided; being able to run the same logic using DuckDB or pandas locally would have been a godsend!

> ## What's next?

Reviewer: Love the post, especially about what values/benefits Kedro on Ibis can offer. It may be valuable to include a section for tradeoffs or current limitations. That way, your audience knows you are not biased.

Reviewer 5: Agree completely - 

Some limitations off the top of my head:

- I think we can highlight that the talent bar for a SQL engineer is much lower. The SQL equivalents of venvs, dependency management and pytest are all trivial in comparison. 
- The dbt / jinja tech debt interest payments may be reasonable depending on your sophistication, pace of development and standards enforcement.
- If you are in an environment where the tech stack has been decided and locked down, flexibility isn't an advantage and using native tooling may be a better bet.

Reviewer 4: @deepyaman I shall leave you to address & resolve the feedback so far. Please ping me here when you're ready for me to take a look at your next draft.

Author: Hmm, good point.

> Some limitations off the top of my head:
>
> * I think we can highlight that the talent bar for a SQL engineer is much lower. The SQL equivalents of venvs, dependency management and pytest are all trivial in comparison.
> * The dbt / jinja tech debt interest payments may be reasonable depending on your sophistication, pace of development and standards enforcement.
> * If you are in an environment where the tech stack has been decided and locked down, flexibility isn't an advantage and using native tooling may be a better bet.

I feel these are less limitations of using Ibis or integrating Kedro and Ibis, and more the Python vs. SQL debate (which I don't want to enter). The goal isn't to convert dbt users to Kedro, but rather to make Kedro users write better pipelines. 

Some ideas that come to mind though, in this vein:

* In general, I think DDL support in Ibis is relatively newer/less fleshed out in Ibis (@cpcloud is this still true? Is this something the sqlglot migration will help significantly in?). Right now, the integration also supports much fewer materializations than dbt does (only `table` and `view`, no `materialized view` or `incremental`).
* On the topic of materializations, dbt also supports `ephemeral`. The Kedro equivalent is `MemoryDataset`, but one thing I didn't do is somehow set default copy mode to assign (like it is for Spark), which would be necessary for it to work correctly. Maybe I can create an issue for this, and link to it?
* dbt has the ability to detect and deploy only modified models; Kedro doesn't have anything automatic like that.

Reviewer 5: Fair ponts - 

> dbt has the ability to detect and deploy only modified models; Kedro doesn't have anything automatic like that.

This is reasonably easy to implement if we make the reasonable assumption that the user is working in a git repo

> If you're familiar with dbt (or even if you examined the Jaffle Shop project discussed above), you'll notice a key functionality that we didn't implement here: [validations](https://docs.getdbt.com/docs/build/validation). Kedro supports data validation through third-party plugins such as [kedro-pandera](https://kedro-pandera.readthedocs.io/en/latest/), and I've recently started work on extending pandera to support validating Ibis tables; look for a follow-up post covering that soon.

Reviewer 2: might be worth pointing out you can also use `pytest` or other standard testing frameworks w/ Ibis

```pycon
>>> import duckdb
>>>
>>> con = duckdb.connect("jaffle_shop.duckdb")
>>> con.sql("SHOW TABLES")
┌───────────────┐
│     name      │
│    varchar    │
├───────────────┤
│ customers     │
│ orders        │
│ raw_customers │
│ raw_orders    │
│ raw_payments  │
│ stg_customers │
│ stg_orders    │
│ stg_payments  │
└───────────────┘

>>> con.sql("SELECT * FROM customers WHERE customer_id = 42")
┌─────────────┬────────────┬───────────┬─────────────┬───────────────────┬──────────────────┬─────────────────────────┐
│ customer_id │ first_name │ last_name │ first_order │ most_recent_order │ number_of_orders │ customer_lifetime_value │
│    int64    │  varchar   │  varchar  │    date     │       date        │      int64       │         double          │
├─────────────┼────────────┼───────────┼─────────────┼───────────────────┼──────────────────┼─────────────────────────┤
│          42 │ Diana      │ S.        │ 2018-02-04  │ 2018-03-12        │                2 │                    27.0 │
└─────────────┴────────────┴───────────┴─────────────┴───────────────────┴──────────────────┴─────────────────────────┘

```

> View the pipeline using Kedro-Viz with `kedro viz run`

Author: ```suggestion
View the pipeline using Kedro-Viz with `kedro viz run`:
```
@stichbury Not sure if the colon was intentionally removed or not.

## What's next?

If you're familiar with dbt (or even if you examined the Jaffle Shop project discussed above), you'll notice a key functionality that we didn't implement here: [validations](https://docs.getdbt.com/docs/build/validation). Kedro natively integrates with pytest for unit testing, which plays well for verifying the correctness of transformations developed in Ibis.  Kedro also supports data validation through third-party plugins such as [kedro-pandera](https://kedro-pandera.readthedocs.io/en/latest/), and I've recently started work on extending pandera to support validating Ibis tables; look for a follow-up post covering that soon.

Ibis supports a subset of DDL operations, which means dbt's `incremental` and `materialized view` materialisations currently don't have counterparts yet. Some Ibis backends have explored exposing materialised views. While not explicitly covered above, the `ephemeral` materialisation equates to Kedro's `MemoryDataset`.

> Finally, dbt offers enhanced deployment functionality, such as the ability to detect and deploy only modified models; Kedro does not trivially allow detecting changes like this.

Reviewer 4: ```suggestion
Finally, dbt offers enhanced deployment functionality, such as the ability to detect and deploy only modified models; it's less straightforward to detect changes with Kedro.
```

Reviewer 4: I'm not massively keen on "allow" unless it's a security/permissions kind of thing. How's this change for you?

Author: ```suggestion
Finally, dbt offers enhanced deployment functionality, like the ability to detect and deploy only modified models; it's less straightforward to detect such changes with Kedro.
```
Seems good! I added a "such" to try to be more clear around what changes we're talking about, but not sure if that's necessary? (Feel free to commit whichever makes more sense to you.)
